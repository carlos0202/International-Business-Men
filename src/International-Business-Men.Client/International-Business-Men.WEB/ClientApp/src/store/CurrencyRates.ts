import { Action, Reducer } from 'redux';
import { AppThunkAction } from './';
import { fillCurrencyTable } from '../utils/transformations';

// -----------------
// STATE - This defines the type of data maintained in the Redux store.

export interface CurrencyRatesState {
    isLoading: boolean;
    currencyRates: CurrencyRate[];
}

export interface CurrencyRate {
    from: string;
    to: string;
    rate: number;
}

export interface ResponseModel<T> {
    statusCode: number,
    statusMessage: string,
    result: T[];
}

// -----------------
// ACTIONS - These are serializable (hence replayable) descriptions of state transitions.
// They do not themselves have any side-effects; they just describe something that is going to happen.

interface RequestCurrencyRatesAction {
    type: 'REQUEST_CURRENCY_RATES';
}

interface ReceiveCurrencyRatesAction {
    type: 'RECEIVE_CURRENCY_RATES';
    currencyRates: CurrencyRate[];
}

// Declare a 'discriminated union' type. This guarantees that all references to 'type' properties contain one of the
// declared type strings (and not any other arbitrary string).
type KnownAction = RequestCurrencyRatesAction | ReceiveCurrencyRatesAction;

// ----------------
// ACTION CREATORS - These are functions exposed to UI components that will trigger a state transition.
// They don't directly mutate state, but they can have external side-effects (such as loading data).

export const actionCreators = {
    requestCurrencyRates: (): AppThunkAction<KnownAction> => (dispatch, getState) => {
        // Only load data if it's something we don't already have (and are not already loading)
        const appState = getState();

        if (appState && appState.currencyRates) {
            fetch(`${process.env.REACT_APP_API_URL}/CurrencyRates`)
                .then(response => response.json() as Promise<ResponseModel<CurrencyRate>>)
                .then(data => {
                    const rates = fillCurrencyTable(data.result);
                    dispatch({ type: 'RECEIVE_CURRENCY_RATES', currencyRates: rates });
                });

            dispatch({ type: 'REQUEST_CURRENCY_RATES' });
        }
    }
};

// ----------------
// REDUCER - For a given state and action, returns the new state. To support time travel, this must not mutate the old state.

const unloadedState: CurrencyRatesState = { currencyRates: [], isLoading: false };

export const reducer: Reducer<CurrencyRatesState> = (state: CurrencyRatesState | undefined, incomingAction: Action): CurrencyRatesState => {
    if (state === undefined) {
        return unloadedState;
    }

    const action = incomingAction as KnownAction;
    switch (action.type) {
        case 'REQUEST_CURRENCY_RATES':
            return {
                currencyRates: state.currencyRates,
                isLoading: true
            };
        case 'RECEIVE_CURRENCY_RATES':
            if (action.currencyRates) {
                return {
                    currencyRates: action.currencyRates,
                    isLoading: false
                };
            }
    }

    return state;
};
