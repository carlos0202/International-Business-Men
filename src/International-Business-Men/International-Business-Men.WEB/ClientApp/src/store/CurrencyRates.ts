import { Action, Reducer } from 'redux';
import { AppThunkAction } from './';

// -----------------
// STATE - This defines the type of data maintained in the Redux store.

export interface CurrencyRatesState {
    isLoading: boolean;
    startDateIndex?: number;
    currencyRates: CurrencyRate[];
}

export interface CurrencyRate {
    from: string;
    to: string;
    rate: number;
}

// -----------------
// ACTIONS - These are serializable (hence replayable) descriptions of state transitions.
// They do not themselves have any side-effects; they just describe something that is going to happen.

interface RequestCurrencyRatesAction {
    type: 'REQUEST_CURRENCY_RATES';
}

interface ReceiveCurrencyRatesAction {
    type: 'RECEIVE_CURRENCY_RATES';
    currencyRates: CurrencyRate[];
}

// Declare a 'discriminated union' type. This guarantees that all references to 'type' properties contain one of the
// declared type strings (and not any other arbitrary string).
type KnownAction = RequestCurrencyRatesAction | ReceiveCurrencyRatesAction;

// ----------------
// ACTION CREATORS - These are functions exposed to UI components that will trigger a state transition.
// They don't directly mutate state, but they can have external side-effects (such as loading data).

export const actionCreators = {
    requestCurrencyRates: (): AppThunkAction<KnownAction> => (dispatch, getState) => {
        // Only load data if it's something we don't already have (and are not already loading)
        const appState = getState();
        console.log(`${process.env.REACT_APP_API_URL}`);
        if (appState && appState.CurrencyRates) {
            fetch(`${process.env.REACT_APP_API_URL}/api/CurrencyRates`)
                .then(response => response.json() as Promise<CurrencyRate[]>)
                .then(data => {
                    dispatch({ type: 'RECEIVE_CURRENCY_RATES', currencyRates: data });
                });

            dispatch({ type: 'REQUEST_CURRENCY_RATES' });
        }
    }
};

// ----------------
// REDUCER - For a given state and action, returns the new state. To support time travel, this must not mutate the old state.

const unloadedState: CurrencyRatesState = { currencyRates: [], isLoading: false };

export const reducer: Reducer<CurrencyRatesState> = (state: CurrencyRatesState | undefined, incomingAction: Action): CurrencyRatesState => {
    if (state === undefined) {
        return unloadedState;
    }

    const action = incomingAction as KnownAction;
    switch (action.type) {
        case 'REQUEST_CURRENCY_RATES':
            return {
                currencyRates: state.currencyRates,
                isLoading: true
            };
        case 'RECEIVE_CURRENCY_RATES':
            // Only accept the incoming data if it matches the most recent request. This ensures we correctly
            // handle out-of-order responses.
            if (action.currencyRates) {
                return {
                    currencyRates: action.currencyRates,
                    isLoading: false
                };
            }
    }

    return state;
};
