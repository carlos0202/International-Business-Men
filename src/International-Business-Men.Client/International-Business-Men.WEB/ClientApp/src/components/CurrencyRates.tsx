import * as React from 'react';
import { connect } from 'react-redux';
import { ApplicationState } from '../store';
import * as CurrencyRatesStore from '../store/CurrencyRates';

// At runtime, Redux will merge together...
type CurrencyRateProps =
    CurrencyRatesStore.CurrencyRatesState // ... state we've requested from the Redux store
    & typeof CurrencyRatesStore.actionCreators; // ... plus action creators we've requested


class CurrencyRates extends React.PureComponent<CurrencyRateProps> {
    // This method is called when the component is first added to the document
    public componentDidMount() {
        this.ensureDataFetched();
    }

    public render() {
        return (
            <React.Fragment>
                <h1 id="tabelLabel">Conversiones</h1>
                <p>Esta es la tabla de conversi&oacute;n entre monedas soportada por el proveedor de datos actual.</p>
                <br />
                {this.renderCurrencyRatesTable()}
            </React.Fragment>
        );
    }

    private ensureDataFetched() {
        this.props.requestCurrencyRates();
    }

    private renderCurrencyRatesTable() {
        return (
            <table className='table table-striped' aria-labelledby="tabelLabel">
                <thead>
                    <tr>
                        <th>Moneda Fuente</th>
                        <th>Moneda Destino</th>
                        <th>Valor de Conversi&oacute;n</th>
                    </tr>
                </thead>
                <tbody>
                    {this.props.currencyRates.map((currencyRate: CurrencyRatesStore.CurrencyRate) =>
                        <tr key={`${currencyRate.from}-${currencyRate.to}`}>
                            <td>{currencyRate.from}</td>
                            <td>{currencyRate.to}</td>
                            <td>{currencyRate.rate}</td>
                        </tr>
                    )}
                </tbody>
            </table>
        );
    }
}

export default connect(
    (state: ApplicationState) => state.currencyRates, // Selects which state properties are merged into the component's props
    CurrencyRatesStore.actionCreators // Selects which action creators are merged into the component's props
)(CurrencyRates as any);
