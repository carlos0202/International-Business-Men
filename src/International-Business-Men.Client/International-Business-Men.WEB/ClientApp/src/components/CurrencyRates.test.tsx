import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import { CurrencyRates } from './CurrencyRates';
import configureStore from 'redux-mock-store';
import { currencyRates } from '../../test/setup/test-data.json';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('', () => {
    let store;
    let component;
    let isLoading = false;

    beforeEach(() => {
        store = mockStore({
            currencyRates, isLoading
        });
        component = renderer.create(
            <Provider store={store}>
                <CurrencyRates {...store} currencyRates={currencyRates}
                    isLoading={isLoading} requestCurrencyRates={jest.fn()} />
            </Provider>
        );
    });

    it('renders CurrencyRates component witouth crashing and to match snapshot', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });
});


