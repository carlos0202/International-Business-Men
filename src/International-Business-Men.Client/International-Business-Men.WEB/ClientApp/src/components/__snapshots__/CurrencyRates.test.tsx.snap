// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CurrencyRates component witouth crashing and to match snapshot 1`] = `
<div
  className="card card-body"
>
  <div
    className="card-title"
  >
    <h1
      id="tabelLabel"
    >
      Conversiones
    </h1>
  </div>
  <br />
  <div>
    <div
      className="table-responsive-md"
    >
      <table
        aria-labelledby="currencyTableLabel"
        className="table table-striped table-bordered table-hover"
      >
        <thead
          className="thead-dark"
        >
          <tr>
            <th>
              Moneda Fuente
            </th>
            <th>
              Moneda Destino
            </th>
            <th>
              Valor de Conversi√≥n
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              EUR
            </td>
            <td>
              USD
            </td>
            <td>
              1.359
            </td>
          </tr>
          <tr>
            <td>
              CAD
            </td>
            <td>
              EUR
            </td>
            <td>
              0.732
            </td>
          </tr>
          <tr>
            <td>
              USD
            </td>
            <td>
              EUR
            </td>
            <td>
              0.736
            </td>
          </tr>
          <tr>
            <td>
              EUR
            </td>
            <td>
              CAD
            </td>
            <td>
              1.366
            </td>
          </tr>
          <tr>
            <td>
              USD
            </td>
            <td>
              AUD
            </td>
            <td>
              1.45
            </td>
          </tr>
          <tr>
            <td>
              AUD
            </td>
            <td>
              USD
            </td>
            <td>
              0.69
            </td>
          </tr>
          <tr>
            <td>
              CAD
            </td>
            <td>
              AUD
            </td>
            <td>
              1.96
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
